File: 491. Non-decreasing Subsequences.cpp
class Solution {
public:
    void backtrack(vector<int>& nums, int index, int n, vector<vector<int>>& arr,
                   vector<int>& currArr, set<vector<int>>& hs) {
        if (index >= n) {
            if (currArr.size() >= 2 && hs.find(currArr) == hs.end()) {
                arr.push_back(currArr);
                hs.insert(currArr);
            }
            return;
        }

        int size = currArr.size();
        if (size >= 1) {
            if (currArr[size - 1] <= nums[index]) {
                currArr.push_back(nums[index]);
                backtrack(nums, index + 1, n, arr, currArr, hs);
                currArr.pop_back();
            }
            backtrack(nums, index + 1, n, arr, currArr, hs);
        } else {
            currArr.push_back(nums[index]);
            backtrack(nums, index + 1, n, arr, currArr, hs);
            currArr.pop_back();
            backtrack(nums, index + 1, n, arr, currArr, hs);
        }
    }

    vector<vector<int>> findSubsequences(vector<int>& nums) {
        vector<vector<int>> arr;
        vector<int> currArr;
        set<vector<int>> hs;
        backtrack(nums, 0, nums.size(), arr, currArr, hs);
        return arr;
    }
};

File: 491. Non-decreasing Subsequences.java
class Solution {
    public void backtrack(int[] nums, int index, int n, List<List<Integer>> arr, ArrayList<Integer> currArr, HashSet<ArrayList<Integer>> hs){
        if(index >= n){
            if(currArr.size() >= 2 && !hs.contains(currArr)){
                arr.add(new ArrayList<Integer>(currArr));
                hs.add(new ArrayList<Integer>(currArr));
            }
            return;
        }

        int size = currArr.size();
        if(size >= 1){
            if(currArr.get(size - 1) <= nums[index]){
                currArr.add(nums[index]);
                backtrack(nums, index + 1, n, arr, currArr, hs);
                currArr.remove(size);
            }
            backtrack(nums, index + 1, n, arr, currArr, hs);
        }
        else{
            currArr.add(nums[index]);
            backtrack(nums, index + 1, n, arr, currArr, hs);
            currArr.remove(0);
            backtrack(nums, index + 1, n, arr, currArr, hs);
        }
    }
    public List<List<Integer>> findSubsequences(int[] nums) {
        List<List<Integer>> arr = new ArrayList<>();
        HashSet<ArrayList<Integer>> hs = new HashSet<>();
        backtrack(nums, 0, nums.length, arr, new ArrayList<Integer>(), hs);

        return arr;
    }
}